### ISSUE TITLE
Metadata syntax changes

### ISSUE THREAD

**Issue by WrinklyNinja** [loot/loot#601]
LOOT v0.10.0 will include metadata changes, so it's a good opportunity to make a few other changes to improve the syntax:

- [x] Rename the `str` key in the message content data structure to `text`, to make its purpose clearer.
- [x] Reimplement `priority` so that it uses the `int8_t` type, giving it a range of values of +/- 127, and denoting global priority using a separate field, `global_priority`. The former change is because the current range of +/- 1,000,000 is overkill and means there's more digits anyone editing needs to keep track of, and the latter is because the modulo on the current field can be confusing and is at least non-obvious. The `global_priority` field overrides the `priority` field in comparisons like so:

    ```
    if (a.global_priority != b.global_priority)
       compare_global_priorities(a, b)
    else
       compare_priorities(a, b)
    ```
- [x] Require path separators used in condition regex file paths to be `/`, not `\`.

**Comment by Freso** [comment]
Will we ever need to make some global `priority`s "more global" than others? This is possible with the current system (e.g., -2000000 vs. -1000000), even if it is not currently used, and might one day be something we want to do again. Maybe make `priority` and `global_priority` both `short`s? (Or maybe there's a smaller integer type that could be used for the global variant?)

**Comment by Freso** [comment]
One benefit of making both integers (instead of one bool and one short) is that it'd be possible to sort e.g., official DLCs using only `global_priority` and not have to use `priority` at all (though for these, you'd probably still want to use both).

**Comment by WrinklyNinja** [comment]
I've decided to restrict the `priority` type to `signed char` rather than `short` (so it's +/- 127), because a `short` is still overkill.

As for some priorities being "more global" than others, I can't think how that would work. Using `global_priority` in place of `priority` is a good idea though, I can make the parser fail if both are provided. I've updated the issue OP with that.

**Comment by Freso** [comment]
I wouldn't make the parser fail if both are provided. If ModA.esp and ModB.esp both have `global_priority: -10` and ModA.esp has `priority: 5` and ModB.esp has `priority: -5`, then ModB.esp should sort before ModA.esp (assuming no master or `after` records override the priority).

**Comment by WrinklyNinja** [comment]
> I wouldn't make the parser fail if both are provided. If ModA.esp and ModB.esp both have global_priority: -10 and ModA.esp has priority: 5 and ModB.esp has priority: -5, then ModB.esp should sort before ModA.esp (assuming no master or after records override the priority).

Yeah, I just realised that a regex entry could cause a plugin to have both, and that wouldn't be caught at parse-time.

However, what if A and B have different priorities and different global priorities, which would take precedence? It's not obvious to me, so I'm thinking for simplicity's sake it may be better to treat `global_priority` as a boolean modifier as originally planned. It's more verbose, but clearer to the reader and simpler code.

**Comment by Freso** [comment]
> However, what if A and B have different priorities and different global priorities, which would take precedence?

The `global_priority` is obviously the one that takes precedence, exactly like now. So if ModA.esp has `global_priority: 5; priority: -5` and ModB.esp has `global_priority: -5; priority: 5`, ModB.esp is sorted first. Exactly like ModA.esp with `priority: 5000095` is sorted after ModB.esp with `priority: -5000105` now.

**Comment by WrinklyNinja** [comment]
So the logic would be as below then?

```
if (a.global_priority != b.global_priority)
  compare_global_priorities(a, b)
else
  compare_priorities(a, b)
```

**Comment by Freso** [comment]
Yes. (I have no idea how LOOT currently deals with the global priorities, but `global_priority` should affect plugin location in the list relative to all other plugins, not just the ones they conflict with. No idea if this has any impact on the logic described.)

**Comment by WrinklyNinja** [comment]
OK, that sounds good then.

**Comment by WrinklyNinja** [comment]
This is now done in the `masterlist-syntax-changes` branch, but I'll leave the issue open until the branch is merged. A couple of screenshots for the new global priority value UI:

![screenshot 54 -item](https://cloud.githubusercontent.com/assets/2430738/16886467/6830f9f0-4acc-11e6-9581-dbb601883165.png)
Sidebar item in edit mode - hovering over each icon/value will display "Global priority" and "Priority" tooltips for the left and right values respectively.


![screenshot 54](https://cloud.githubusercontent.com/assets/2430738/16886465/666d7120-4acc-11e6-8681-af9d3aefcdd1.png)
Editor

**Comment by Freso** [comment]
Perhaps put out a 0.9.2 release to fix the Oblivion thing before merging the branch?
Will you be providing a snapshot build of this before merging? Are you going to do an automated conversion of current masterlist files for the priorities, or should we go through them manually?

**Comment by WrinklyNinja** [comment]
Yeah, I'll probably do a 0.9.2 release soon, there have been a bunch of fixes since 0.9.1. There's already a snapshot of the branch available, and yes, I'll convert the masterlists when I create the `v0.10` branches for them.

**Comment by Freso** [comment]
How are you going to do the conversion? `global_priority = x // 10^6` and `priority = (x % 10^6) / (10^6 - 1) * 128`?

**Comment by WrinklyNinja** [comment]
It'll be manual - I'll have a look at what values are currently being used, and map them to the new range so that each unique value used stays unique.

**Comment by pStyl3** [comment]
Whenever I try to set a global priority value in the editor, a value of n+48 is saved. So if I want a GP to be 1 it is safed as 49, 2 as 50 and so on.

Also, -1 will produce the "Error code: -1; Failed to apply plugin metadata. Details: bad conversion" error.

**Comment by WrinklyNinja** [comment]
Ah, the values were being interpreted as characters, fixed.

**Comment by pStyl3** [comment]
Hmm, can't we keep the old star icon for global priority I imagine your problem was finding another suitable icon for the normal priority .. how about something like [this](https://abload.de/img/iconsr8ure.jpg), an unfilled star icon for priority and a full star icon for global priority (as it was before)?

**Comment by WrinklyNinja** [comment]
Yes, filled/empty star icons could be used.

**Comment by WrinklyNinja** [comment]
OK, I changed to filled/empty stars because it does look much better.

**Comment by WrinklyNinja** [comment]
I've added a new change to this issue: file path separators in regex conditions will have to be `/`, not `\`, as the latter is also a special character in regex and the current flexibility allows for some confusion. I don't know if this actually affects anything in the masterlists, but it will be minor if it does.

**Comment by Freso** [comment]
`/` as a file path separator is the more natural choice too! 😜

We do have some SKSE/F4SE script checks that use paths (both e.g., `../??se_launcher.exe` to check for ??SE, and also to check for whether the *SE plugins are there in the first place. We may also use them to check for some things that are not .esp's (e.g., I may add an FO4 masterlist check for `interface/DEF_CONF/DEF_INV_TAGS.xml` as a check for whether DEF_INV is in use). But as you said, it's only about a handful of cases per masterlist that are likely to be affected, so it should work out fine.

**Comment by WrinklyNinja** [comment]
Merged in 4baad635b1d892c1b89e59eb385c8f34d398e8ff
